QS: Collection Batch Reveal (#1434)
smartcontractkit/documentation: src/pages/chainlink-automation/tutorials/nft-quickstart.mdx

---
layout: ../../../layouts/MainLayout.astro
section: automation
date: Last Modified
title: "Batch NFT Reveal"
whatsnext: { "FAQs": "/chainlink-automation/faqs" }
---

import { Aside, CodeSample, ClickToZoom } from "@components"
import { Tabs } from "@components/Tabs"

Using [Chainlink VRF](https://docs.chain.link/vrf/v2/introduction/) in generative art NFT collections is de-facto the standard approach for getting provably random source in smart contracts. By batching the reveal process, instead of making VRF calls for each NFT we can save cost up to 100x (in a collection of 10,000 with batch size of 100).

> This template uses the [Chainlink Automation](https://automation.chain.link) decentralized network to automate the reveal process so you don't have to stand up and maintain an inhouse system for automation.

Reviewer: ```suggestion
This template uses the [Chainlink Automation](https://automation.chain.link) decentralized network to automate the reveal process so you don't have to stand up and maintain an in-house system for automation.
```

{/* prettier-ignore */}
<div class="remix-callout">
  <a href="https://github.com/smartcontractkit/chainlink-automation-templates/tree/main/batch-nft-reveal">See the code on GitHub</a>
</div>

<Aside type="caution" title="Disclaimer">
  This tutorial is an example to help you learn how to integrate Chainlink services and integrate them into your own
  dApps. This template is provided "AS IS" without warranties of any kind, has not been audited, and may be missing key
  checks or error handling to make the usage of the product more clear. Do not use the code in this example in a
  production environment without completing your own audits and application of best practices.

> </Aside>

Reviewer: Reminder to update this disclaimer

Author: Resolved

## Overview

> In this tutorial, you will use a [demo user interface](https://automation.chainlink-demo.app) to learn how to use Chainlink Automation to automate batch reveal of NFT mints. The demo includes setting up a Chainlink VRF subscription, funding it, automatically generating a smart contract that will then be used by both VRF (as a consumer) and in Automation (as the target contract). VRF is used for randomizing the metadata for the NFTs and Automation is used to set trigger conditions for when the NFT metadata is revealed via batch size and/or time interval parameters

Reviewer: ```suggestion
In this tutorial, you will use a [demo user interface](https://automation.chainlink-demo.app) that helps you learn how to automate batch reveal of NFT mints, using Chainlink Automation and VRF. VRF is used to randomize the metadata for the NFTs, and Automation is used to set trigger conditions for when the NFT metadata is revealed via batch size and/or time interval parameters.

The demo includes:

- Setting up and funding a Chainlink VRF subscription
- Automatically generating a smart contract that will then be used by both VRF (as a consumer) and in Automation (as the target contract). 
```

## Before you begin

Before you start this tutorial, complete the following items:

- Set up a cryptocurrency wallet
- To deploy this contract on testnets, ensure the deployer account has testnet ERC-677 LINK. Use the [LINK faucet](https://faucets.chain.link/) to retrieve testnet LINK.

## Run steps

1. Go to the [demo app](https://automation.chainlink-demo.app/) and connect to your wallet with the right network. Then click on the **Get Started** button.

   <ClickToZoom
     src="/images/automation/tutorials/batch-nft1.png"
     alt="Batch NFT Quickstart Step 1"
     style="max-width: 70%;"
   />

> 1. Enter the basic information regarding your NFT collection. You can hover over the question mark icon to learn more about each field. Notice that you'll need a VRF Subscription ID.

Reviewer: ```suggestion
1. Enter the basic information regarding your NFT collection. You can hover over the question mark icon to learn more about each field.
```
Pointing out at this step that I'll need a VRF subscription ID makes it sound like I need it for this step. This was mildly confusing until I realized we'd be setting it up in the next step.

   <ClickToZoom
     src="/images/automation/tutorials/batch-nft2.png"
     alt="Batch NFT Quickstart Step 2"
     style="max-width: 70%;"
   />

1. At this stage, you will need to create a VRF subscription and fund it with LINK by visiting the [Chainlink VRF Subscription Manager](https://vrf.chain.link/). Follow the steps outlined in the interface to create a subscription. You will receive a confirmation message upon successful creation.

> 1. Add funds to your subscription. Copy the resulting subscription ID to your clipboard.

Reviewer: Include a good starting point for how much LINK someone might need to fund their VRF subscription. (Although it is shown in the screenshot, also add it to the text here for more clarity and accessibility.)

Author: Resolved

<ClickToZoom
  src="/images/automation/tutorials/batch-nft6.png"
  alt="Batch NFT Quickstart Step 6"
  style="max-width: 70%;"
/>

1. The next step is to register a new Upkeep using Chainlink Automation. Visit the [Chainlink Automation App](https://automation.chain.link/) to get started. Click the button **Register New Upkeep** and connect your wallet.

1. You will be directed to a screen where you will choose your Trigger from a dropdown menu. Select **Custom logic**. Paste your address into the _Target contract address_ field. Then, click **Next**.

1. You will be directed to a form to fill out more information on your upkeep. Fill out these details and set the gas limit to 200000 to ensure the Chainlink Automation network can perform the upkeep when the trigger conditions are met. Once all the information is filled in, click **Register Upkeep**. Wait until the upkeep has been registered and click **View Upkeep**. You'll be able to see all the details of your Upkeep.

> 1. Navigate back to the demo app page. Check off both boxes and click **View Collection**. You screen will look similar to this.

Reviewer: ```suggestion
1. Navigate back to the demo app page. Check off both boxes and click **View Collection**. Your screen will look similar to this:
```

<ClickToZoom
  src="/images/automation/tutorials/batch-nft7.png"
  alt="Batch NFT Quickstart Step 7"
  style="max-width: 70%;"
/>

> 1. Mint NFTs until the trigger condition is met. At that point, the demo app will display the message "Pending Batch Reveal".The Chainlink Automation upkeep eill be triggered automatically. This will internally make a call to Chainlink VRF for randomness and assign the random traits to the tokens for the reveal.

Reviewer: ```suggestion
1. Mint NFTs until the trigger condition is met. At that point, the demo app will display the message "Pending Batch Reveal". The Chainlink Automation upkeep will be triggered automatically. This will internally make a call to Chainlink VRF for randomness and assign the random traits to the tokens for the reveal.
```

> 2. If the reveal is a success, your screen will look similar to this. You have no completed the tutorial.

Reviewer: ```suggestion
1. If the reveal is a success, your screen will look similar to this. You have now completed the tutorial.
```

  provided “AS IS” and “AS AVAILABLE” without warranties of any kind, has not been audited, and may be missing key
  checks or error handling to make the usage of the product more clear. Do not use the code in this example in a
  production environment without completing your own audits and application of best practices. Neither Chainlink Labs,
  the Chainlink Foundation, nor Chainlink node operators are responsible for unintended outputs that are generated due
  to errors in code.
</Aside>

## Overview

In this tutorial, you will use a [demo user interface](https://automation.chainlink-demo.app) that helps you learn how to automate batch reveal of NFT mints, using Chainlink Automation and VRF. VRF is used to randomize the metadata for the NFTs, and Automation is used to set trigger conditions for when the NFT metadata is revealed via batch size and/or time interval parameters.

The demo includes:

- Setting up and funding a Chainlink VRF subscription
- Automatically generating a smart contract that will then be used by both VRF (as a consumer) and in Automation (as the target contract).

## Before you begin

Before you start this tutorial, complete the following items:

> - Set up a cryptocurrency wallet
> - To deploy this contract on testnets, ensure the deployer account has testnet ERC-677 LINK. Use the [LINK faucet](https://faucets.chain.link/) to retrieve testnet LINK.

Reviewer 2: ```suggestion
- If you are new to smart contract development, learn how to [Deploy Your First Smart Contract](/getting-started/deploy-your-first-contract).
- Set up a cryptocurrency wallet
- To deploy this contract on testnets, ensure the deployer account has testnet ERC-677 LINK. Use the [LINK faucet](https://faucets.chain.link/) to retrieve testnet LINK.
```

---
layout: ../../../layouts/MainLayout.astro
section: automation
date: Last Modified
title: "Batch NFT Reveal"
whatsnext: { "FAQs": "/chainlink-automation/faqs" }
---

import { Aside, CodeSample, ClickToZoom } from "@components"
import { Tabs } from "@components/Tabs"

> Using [Chainlink VRF](https://docs.chain.link/vrf/v2/introduction/) in generative art NFT collections is de-facto the standard approach for getting provably random source in smart contracts. By batching the reveal process, instead of making VRF calls for each NFT we can save cost up to 100x (in a collection of 10,000 with batch size of 100).

Reviewer 2: ```suggestion
Using [Chainlink VRF](https://docs.chain.link/vrf/v2/introduction/) in generative art NFT collections is the standard approach for getting provably random source in smart contracts. By batching the reveal process, instead of making VRF calls for each NFT we can save cost up to 100x (in a collection of 10,000 with batch size of 100).
```
